### PR TITLE
Versioning 3 merge

### PR DESCRIPTION
**Pull request by dnr**
**Do not squash! This is intended to be merged with a real merge commit.**

**Need to edit the first commit with final merged api-go commit after api is merged.**
**This won't build yet until we merge api properly.**

## What changed?
Cleaner version of versioning-3 merge

### CODE REVIEW

File: service/matching/deployment_util.go
Comment (alexshtin) [inline service/matching/deployment_util.go:37]: I know there is no consistency about case and punctuation in serviceerrors but can at least these two lines be consistent:
```suggestion
	errDeploymentsNotAllowed = serviceerror.NewPermissionDenied("deployments are disabled on this namespace", "")
	errMissingDeployment     = serviceerror.NewInvalidArgument("missing Deployment field")
```

**Review by alexshtin — APPROVED** [review]

File: service/history/workflow/mutable_state.go
Comment (ShahabT) [reply service/history/workflow/mutable_state.go:443]: Well... one way to see it is `DeploymentTransision` is not about deployments it's about workflow execution transitioning between deployments.

File: service/history/ndc/workflow_resetter_test.go
Comment (ShahabT) [reply service/history/ndc/workflow_resetter_test.go:997]: No this is needed to get one test passing with latest API changes, because of api-go pipeline being broken, non-versioning changes also crawled into the api-go commit :( (the same line change is in https://github.com/temporalio/temporal/pull/6888)

File: service/matching/deployment_util.go
Comment (carlydf) [inline service/matching/deployment_util.go:37]:
```suggestion
	errDeploymentsNotAllowed = serviceerror.NewPermissionDenied("Deployments are disabled on this namespace", "")
	errMissingDeployment     = serviceerror.NewInvalidArgument("missing Deployment")
```

Comment (carlydf) [reply service/matching/deployment_util.go:37]: I made myself a note to fix the "dD" after merge to main, so don't let this block the merge

Comment (ShahabT) [reply service/matching/deployment_util.go:37]: If the tests pass, Let's fix the typo in a subsequent PR. (I wanted to start the message with small letters, per Go conventions.)

**Pull request merged**
